### PR TITLE
An alternative EffectiveConfig

### PR DESCRIPTION
**Pull request by cnsnyder**
This PR adds the following classes:
- EffectiveConfig: A container object that represents the effective configuration that virt-who is currently running with. It maintains the hierarchy of configuration sources applied to all sections. It has dictionary like access. It contains ConfigSection objects.

- ConfigSection: This is the base class for each type of section we might have in virt-who configuration. It contains the logic to validate a section of the configuration as well as serves as a collection of that configuration. It also has dictionary like access. It also has a state variable which it updates every time the validate method is run. The validate method returns a list of ("log_level", "message"), which are meant to be any sort of validation output that the section deems necessary. It is also the responsibility of the validate method to deserialize (if necessary) any of the parameters given it (so that anything using that ConfigSection in the code base need not worry about that).

- GlobalConfigSection: A subclass of ConfigSection. Implements the validation method and fills in some DEFAULTS. Note not all configuration sections will be validated the same way so this might not be the very best example of how to write a ConfigSection for a virt backend. it should be fairly similar though.

The remainder of the changes were to get the EffectiveConfig in at least some use throughout the code base. The EffectiveConfig has essentially replaced the "options" objects seen throughout the code base. Other updates were to mainly to fix broken tests.

### CODE REVIEW

**Comment by cnsnyder** [comment]
Reviewer: Please merge only #81 OR #82

**Comment by cnsnyder** [comment]
[[13] [CONFIG_EPIC] Create ConfigHierarchy Class](https://trello.com/c/RF32mpFs/451-13-configepic-create-confighierarchy-class)

File: virtwho/config.py [hunk 1]
```diff
@@ -18,21 +18,30 @@
...
 import os
-
 from ConfigParser import SafeConfigParser, NoOptionError, Error, MissingSectionHeaderError
-from virtwho import DefaultInterval
+from virtwho import log
 from password import Password
 from binascii import unhexlify
 import hashlib
 import json
+import six
```
Comment (jirihnidek) [inline virtwho/config.py]: Module `six` is not used. I would not mix this refactoring with effort to add support for Python 3.

File: virtwho/config.py [hunk 2]
```diff
@@ -18,21 +18,30 @@
...
 
-VIRTWHO_CONF_DIR = "/etc/virt-who.d/"
+# Module-level logger
+logger = log.getLogger('config', queue=False)
+
+VW_CONF_DIR = "/etc/virt-who.d/"
 VIRTWHO_TYPES = ("libvirt", "vdsm", "esx", "rhevm", "hyperv", "fake", "xen")
 VIRTWHO_GENERAL_CONF_PATH = "/etc/virt-who.conf"
-VIRTWHO_GLOBAL_SECTION_NAME = "global"
+VW_GLOBAL = "global"
```
Comment (jirihnidek) [inline virtwho/config.py]: `VW` stands for Volkswagen, right? ;-) ... Seriously, I would use VW in other constants too.

File: virtwho/config.py [hunk 3]
```diff
@@ -327,21 +336,21 @@ class GlobalConfig(GeneralConfig):
...
         'log_per_config'
     )
     INT_OPTIONS = (
         'interval',
     )
 
     @classmethod
-    def fromFile(cls, filename, logger=None):
-        global_config = parseFile(filename, logger=logger).get(VIRTWHO_GLOBAL_SECTION_NAME)
+    def fromFile(cls, filename):
```
Comment (jirihnidek) [inline virtwho/config.py]: I would not use camelCase and I would rename it to: `from_file()`.

File: virtwho/config.py [hunk 4]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    for section in sections:
+        if '__name__' in sections[section]:
+            del sections[section]['__name__']
     return sections
+
+# Helper methods used to validate parameters given to virt-who
+def str_to_bool(value):
+    if isinstance(value, bool):
+        return value
+    if isinstance(value, str):
```
Comment (jirihnidek) [inline virtwho/config.py]: To advocate importing `six` module you can use here: `if isinstance(value, six.string_types):

File: virtwho/config.py [hunk 5]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
-    for section in parser.sections():
-        try:
-            sections[section] = getOptions(section, parser)
-        except NoOptionError:
-            sections[section] = {}
-    return sections
-
-
-def parseFile(filename, logger=None):
+def parseFile(filename):
```
Comment (jirihnidek) [inline virtwho/config.py]: I would use: `def parse_file(filename)`

File: virtwho/config.py [hunk 6]
```diff
@@ -299,7 +308,7 @@ def has_options(self, options):
         return True
 
     @classmethod
-    def fromFile(cls, filename, logger):
+    def fromFile(cls, filename):
```
Comment (jirihnidek) [inline virtwho/config.py]: `from_file()`

File: virtwho/config.py [hunk 7]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+        raise TypeError("Value is not a string")
+    if not value:
+        raise ValueError("String is empty")
+    return value
+
+
+def readable(path):
+    if not os.access(path, os.R_OK):
+        raise ValueError("Path '%s' is not accessible" % path)
+    return path
```
Comment (jirihnidek) [inline virtwho/config.py]: When you return value, then it suggests that you do something with `path`. It is not this case.

File: virtwho/config.py [hunk 8]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+
+
+def accessible_file(path):
+    readable(path)
+    if not os.path.isfile(path):
+        raise ValueError("Path '%s' does not appear to be a file" % path)
+    return path
```
Comment (jirihnidek) [inline virtwho/config.py]: Same here.

File: virtwho/config.py [hunk 9]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    if not os.path.isdir(path):
+        raise ValueError("Path '%s' does not appear to be a directory" % path)
+
+
+def empty_or_accessible_files(paths):
+    if not isinstance(paths, list):
+        if not isinstance(paths, str):
+            raise TypeError()
+        if len(paths) == 0:
+            return []
```
Comment (jirihnidek) [inline virtwho/config.py]: I guess this case is solved at line 754 too.

File: virtwho/config.py [hunk 10]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+def accessible_dir(path):
+    readable(path)
+    if not os.path.isdir(path):
+        raise ValueError("Path '%s' does not appear to be a directory" % path)
+
+
+def empty_or_accessible_files(paths):
+    if not isinstance(paths, list):
+        if not isinstance(paths, str):
+            raise TypeError()
```
Comment (jirihnidek) [inline virtwho/config.py]: What about adding some string to this `TypeError()`. Something like: `TypeError('Path has to be list of strings or string')

File: virtwho/config.py [hunk 11]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    return path
+
+
+def accessible_dir(path):
+    readable(path)
+    if not os.path.isdir(path):
+        raise ValueError("Path '%s' does not appear to be a directory" % path)
+
+
+def empty_or_accessible_files(paths):
```
Comment (jirihnidek) [inline virtwho/config.py]: I would add some doc string for this method, because the name is no self-explanatory.

File: virtwho/config.py [hunk 12]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    return paths
+
+
+class ConfigSection(collections.MutableMapping):
+    """
+    This represents a section of configuration for virt-who. The interface that it exposes is 
+    dictionary like. This object maintains a state attribute. The state shows if the configuration
+    section has passed validation, needs validation, or is invalid for some reason.
+    """
+    # The string representation of all default properties and values
```
Comment (jirihnidek) [inline virtwho/config.py]: Add note that real values will be added in child classes.

File: virtwho/config.py [hunk 13]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    DEFAULTS = ()
+    REQUIRED = ()
+
+    __marker = object()
+
+    def __init__(self, section_name, wrapper):
+        self.section_name = section_name
+        self._wrapper = wrapper
+        self.defaults = dict(type(self).DEFAULTS)
```
Comment (jirihnidek) [inline virtwho/config.py]: Why not just `self.DEFAULTS`?

File: virtwho/config.py [hunk 14]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+        # Those properties that have yet to be validated
+        # Any item not in this set but in this ConfigSection is validated
+        self._unvalidated_keys = set()
+        self._invalid_keys = set()
+        # Holds all properties for this section
+        self._values = {}
+        # Add section defaults
+        for key, value in self.defaults.iteritems():
```
Comment (jirihnidek) [inline virtwho/config.py]: Not Python 3 friendly, use `self.defaults.items()`.

File: virtwho/config.py [hunk 15]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+                for key, value in arg:
+                    self._values[key] = value
+        for key, value in kwds.iteritems():
+            self._values[key] = value
+
+    @classmethod
+    def get_defaults(cls):
+        defaults = dict()
+        for key, value in cls.DEFAULTS:
+            defaults[key] = str(value)
```
Comment (jirihnidek) [inline virtwho/config.py]: Why casting to `str`?

File: virtwho/config.py [hunk 16]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+
+    def update(self, *args, **kwds):
+        for arg in args:
+            if getattr(arg, 'keys', None):
+                for key in arg.keys():
+                    self._values[key] = arg[key]
+            else:
+                for key, value in arg:
+                    self._values[key] = value
+        for key, value in kwds.iteritems():
```
Comment (jirihnidek) [inline virtwho/config.py]: Use `kwds.items()` to be compatible with Python 3.

File: virtwho/config.py [hunk 17]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+                return default
+            raise e
+
+    def __str__(self):
+        return_string = "[%s]" % self.section_name
+        for key, value in self.iteritems():
+            return_string += "\n%s=%s" % (key, value)
+        return return_string
+
+    def update(self, *args, **kwds):
```
Comment (jirihnidek) [inline virtwho/config.py]: This method deserves doc string.

File: virtwho/config.py [hunk 18]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+        return self.defaults[key] == self._values[key]
+
+    def has_default(self, key):
+        return key in self.defaults
+
+    def get(self, key, default=__marker):
+        try:
+            return self[key]
+        except KeyError as e:
+            if default is not self.__marker:
```
Comment (jirihnidek) [inline virtwho/config.py]: What? Why? In what case this is not `True`?

File: virtwho/config.py [hunk 19]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+            defaults[key] = str(value)
+        return defaults
+
+    @classmethod
+    def class_for_type(cls, virt_type):
+        clazz = ConfigSection
+        # TODO: Update to use actual subclasses as the are created
+        for subclass in cls.__subclasses__():
+            if subclass.VIRT_TYPE == virt_type:
+                clazz = subclass
```
Comment (jirihnidek) [inline virtwho/config.py]: I would add `break` here.

File: virtwho/config.py [hunk 20]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+                                                   'See man virt-who-config for more info')
+        return result
+
+    def _validate_configs(self):
+        result = None
+        if self.is_default('configs'):
+            self._values['configs'] = []
+        elif isinstance(self['configs'], list):
+            self._values['configs'] = [item for item in self._values['configs'] if isinstance(item,
+                                                                                          str)]
```
Comment (jirihnidek) [inline virtwho/config.py]: I hope it is possible to break the line in better way.

File: virtwho/config.py [hunk 21]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+        elif not isinstance(self._values['configs'], str):
+            result = ('warning', '"configs" must be one or more strings, '
+                                                   'ignoring')
+            self._values['configs'] = []  # Reset to empty list
+        else:
+            self._values['configs'] = self._values['configs'].split(',')
```
Comment (jirihnidek) [inline virtwho/config.py]: I would switch elif-else here to something like this:

```python
elif isinstance(self._values['configs'], str):
    # ...
else:
    # ...
```

File: virtwho/config.py [hunk 22]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+            result = ('warning', '%s cannot be empty, using default' % key)
+        return result
+
+    def validate(self, validation_messages=None):
+        validation_messages = super(GlobalSection, self).validate(validation_messages)
+        to_reset = set()
+        # Validate those keys that need to be validated
+        for key in set(self._unvalidated_keys):
+            error = None
+            # If boolean
```
Comment (jirihnidek) [inline virtwho/config.py]: ?

File: virtwho/config.py [hunk 23]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+            elif key == 'configs':
+                error = self._validate_configs()
+            elif key in ['reporter_id', 'log_file', 'log_dir']:
+                error = self._validate_non_empty_string(key)
+            elif key == 'interval':
+                error = self._validate_interval()
+            else:
+                validation_messages.append(('warning', 'Ignoring unknown configuration option '
+                                                       '"%s"' % key))
+                del self._values[key]
```
Comment (jirihnidek) [inline virtwho/config.py]: I would use `dict` dispatcher in this case.

File: virtwho/config.py [hunk 24]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+            new_section.update(**virt_defaults_section)
+            new_section.update(**values)
+            try:
+                new_section = ConfigSection.from_dict(new_section, section, self)
+            except KeyError:
+                # Missing required attribute
+                pass
+            self._sections[section] = new_section
+
+
```
Comment (jirihnidek) [inline virtwho/config.py]: Too much white space.

File: virtwho/config.py (hunk 18 above)
Comment (cnsnyder) [inline virtwho/config.py]: To allow the default to be None I had to set the optional arg to something other than None.
This way if the user sets default to anything (so long as it is not self.__marker) and the given key is not in self, we'll return default.

File: virtwho/config.py [hunk 25]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+                error = self._validate_interval()
+            else:
+                validation_messages.append(('warning', 'Ignoring unknown configuration option '
+                                                       '"%s"' % key))
+                del self._values[key]
+            if error is not None:
+                validation_messages.append(error)
+                if key not in ['interval']:  # Special cases not reset to default on failure
+                    to_reset.add(key)
+                self._invalid_keys.add(key)
```
Comment (cnsnyder) [inline virtwho/config.py]: Yea I could see that I'll figure out another name. The intention behind it is a set of keys whose values are invalid.

File: virtwho/config.py [hunk 26]
```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
...
+    def __len__(self):
+        return len(self._sections)
+
+    def __getitem__(self, key):
+        return self._sections[key]
+
+    def __contains__(self, item):
+        return item in self._sections
+
+    def __init__(self, env_args, cli_args):
```
Comment (cnsnyder) [inline virtwho/config.py]: Would the preference to be have a module level method which creates an instance of the EffectiveConfig object?

File: virtwho/config.py (hunk 4 above)
Comment (cnsnyder) [inline virtwho/config.py]: Let's comeback to python 3 compatibility once we've looked at the changes necessary.
For now I'll go with the removal of six. (which I added accidentally).

File: virtwho/config.py (hunk 2 above)
Comment (cnsnyder) [inline virtwho/config.py]: Sounds good, done!

**Pull request merged**